README

Swfdec is a decoder/renderer for Macromedia Flash animations.  The
decoding and rendering engine is provided in a library that can
be used by other applications.

This package contains programs to integrate Flash functionality into the GNOME
desktop.  It's main application is swfdec-player, a stand-alone viewer for 
Flash files.  It also contains swfdec-thumbnailer, a program that provides 
screenshots for files to display in the Nautilus file manager


Maintainers:

   see the file MAINTAINERS


Dependencies:

  swfdec (= 0.5.4)
  swfdec-gtk (= 0.5.4)
  gtk (>= 2.12.0)
  GConf2



NEWS

 0.5.5 ("Banks Peninsula Track")

not a lot of new things.
- added translation: pl

 0.5.4 ("Turkish Cycling Federation")

initial release. Contains a thumbnailer and a standalone player.


player/swfdec-player.c
/* Swfdec Player
 * Copyright (C) 2006 Benjamin Otte <otte@gnome.org>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA  02110-1301, USA.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdlib.h>
#include <string.h>
#include <glib/gi18n.h>
#include "swfdec-window.h"

static char *
sanitize_url (const char *s)
{
  if (strstr (s, "://")) {
    return g_strdup (s);
  } else {
    char *dir, *full;
    if (g_path_is_absolute (s))
      return g_strconcat ("file://", s, NULL);
    dir = g_get_current_dir ();
    full = g_strconcat ("file://", dir, G_DIR_SEPARATOR_S, s, NULL);
    g_free (dir);
    return full;
  }
}

int 
main (int argc, char *argv[])
{
  SwfdecWindowSettings settings = { TRUE, TRUE };
  GError *error = NULL;
  char **filenames = NULL;
  SwfdecWindow *window;
  guint i;
  char *s;

  GOptionEntry options[] = {
    { "no-sound", 'n', G_OPTION_FLAG_REVERSE, G_OPTION_ARG_NONE, &settings.sound, N_("don't play sound"), NULL },
    { "paused", 'p', G_OPTION_FLAG_REVERSE, G_OPTION_ARG_NONE, &settings.playing, N_("start player paused"), NULL },
    { G_OPTION_REMAINING, 0, 0, G_OPTION_ARG_STRING_ARRAY, &filenames, NULL, "<FILE> [<FILE> ...]" },
    { NULL }
  };
  GOptionContext *ctx;

  g_set_application_name (_("Swfdec Flash Player"));
  ctx = g_option_context_new ("");
  g_option_context_set_translation_domain (ctx, GETTEXT_PACKAGE);
  g_option_context_add_main_entries (ctx, options, "options");
  g_option_context_add_group (ctx, gtk_get_option_group (TRUE));
  g_option_context_parse (ctx, &argc, &argv, &error);
  g_option_context_free (ctx);

  if (error) {
    g_printerr ("Error parsing command line arguments: %s\n", error->message);
    g_error_free (error);
    return EXIT_FAILURE;
  }

  swfdec_init ();

  if (filenames == NULL) {
    /* open an empty window if no args */
    window = swfdec_window_new (NULL);
    swfdec_window_set_settings (window, &settings);
  } else {
    for (i = 0; filenames[i]; i++) {
      s = sanitize_url (filenames[i]);
      window = swfdec_window_new (s);
      g_free (s);
      swfdec_window_set_settings (window, &settings);
    }
    g_strfreev (filenames);
  }

  gtk_main ();

  return EXIT_SUCCESS;
}



player/swfdec-window.h
/* Swfdec Player
 * Copyright (C) 2007 Benjamin Otte <otte@gnome.org>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA  02110-1301, USA.
 */

#include <gtk/gtk.h>
#include <libswfdec/swfdec.h>
#include <libswfdec-gtk/swfdec-gtk.h>

#ifndef __SWFDEC_WINDOW_H__
#define __SWFDEC_WINDOW_H__


typedef struct _SwfdecWindow SwfdecWindow;
typedef struct _SwfdecWindowSettings SwfdecWindowSettings;
typedef struct _SwfdecWindowClass SwfdecWindowClass;

#define SWFDEC_TYPE_WINDOW                    (swfdec_window_get_type())
#define SWFDEC_IS_WINDOW(obj)                 (G_TYPE_CHECK_INSTANCE_TYPE ((obj), SWFDEC_TYPE_WINDOW))
#define SWFDEC_IS_WINDOW_CLASS(klass)         (G_TYPE_CHECK_CLASS_TYPE ((klass), SWFDEC_TYPE_WINDOW))
#define SWFDEC_WINDOW(obj)                    (G_TYPE_CHECK_INSTANCE_CAST ((obj), SWFDEC_TYPE_WINDOW, SwfdecWindow))
#define SWFDEC_WINDOW_CLASS(klass)            (G_TYPE_CHECK_CLASS_CAST ((klass), SWFDEC_TYPE_WINDOW, SwfdecWindowClass))
#define SWFDEC_WINDOW_GET_CLASS(obj)          (G_TYPE_INSTANCE_GET_CLASS ((obj), SWFDEC_TYPE_WINDOW, SwfdecWindowClass))

struct _SwfdecWindowSettings
{
  gboolean		playing;	/* TRUE if this window should be playing automagically */
  gboolean		sound;		/* TRUE if sund is active */
};

struct _SwfdecWindow
{
  GObject		object;

  gboolean		error;		/* TRUE if we're in error */
  GtkBuilder *		builder;	/* builder instance to load from */
  GtkWidget *		window;		/* the toplevel window */
  SwfdecPlayer *	player;		/* the player we show or NULL if not initialized yet */
  SwfdecLoader *	loader;		/* the loader we use to load the content or NULL if not initialized yet */
  SwfdecWindowSettings	settings;	/* the settings that apply to this window */
};

struct _SwfdecWindowClass
{
  GObjectClass		object_class;
};

SwfdecWindow *	swfdec_window_new		(const char *			url);

gboolean	swfdec_window_set_url		(SwfdecWindow *			window,
						 const char *			url);
void		swfdec_window_error		(SwfdecWindow *			window,
						 const char *			msg);
void		swfdec_window_set_settings	(SwfdecWindow *			window,
						 const SwfdecWindowSettings *	settings);


G_END_DECLS
#endif


player/swfdec-window-handlers.c
/* Swfdec Player
 * Copyright (C) 2007 Benjamin Otte <otte@gnome.org>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA  02110-1301, USA.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <glib/gi18n.h>

#include "swfdec-window.h"

static void
menu_file_open_response (GtkFileChooser *chooser, gint response, SwfdecWindow *window)
{
  if (response == GTK_RESPONSE_ACCEPT) {
    char *target = gtk_file_chooser_get_filename (chooser);
    if (target != NULL) {
      char *url = g_strconcat ("file://", target, NULL);

      if (!swfdec_window_set_url (window, url)) {
	SwfdecWindow *new = swfdec_window_new (url);
	swfdec_window_set_settings (new, &window->settings);
      }
      g_free (url);
      g_free (target);
    }
  }
    
  gtk_widget_destroy (GTK_WIDGET (chooser));
  return;
}

void
menu_file_open (GtkAction *action, SwfdecWindow *window)
{
  GtkWidget *chooser;

  chooser = gtk_file_chooser_dialog_new (_("Select a file to play"),
      GTK_WINDOW (window->window), 
      GTK_FILE_CHOOSER_ACTION_OPEN, 
      GTK_STOCK_CANCEL, GTK_RESPONSE_CANCEL,
      GTK_STOCK_OPEN, GTK_RESPONSE_ACCEPT, NULL);
  gtk_dialog_set_default_response (GTK_DIALOG (chooser), GTK_RESPONSE_ACCEPT);

  g_object_ref (window);
  g_signal_connect (chooser, "response", G_CALLBACK (menu_file_open_response), window);
  g_object_weak_ref (G_OBJECT (chooser), (GWeakNotify) g_object_unref, window);
  gtk_window_present (GTK_WINDOW (chooser));
}

void
menu_file_play (GtkToggleAction *action, SwfdecWindow *window)
{
  GObject *o;

  window->settings.playing = gtk_toggle_action_get_active (action);
  if (window->player) {
    swfdec_gtk_player_set_playing (SWFDEC_GTK_PLAYER (window->player), 
	window->settings.playing);
    o = gtk_builder_get_object (window->builder, "player-widget");
    swfdec_gtk_widget_set_interactive (SWFDEC_GTK_WIDGET (o), window->settings.playing);
  }
}

void
menu_file_mute (GtkToggleAction *action, SwfdecWindow *window)
{
  window->settings.sound = !gtk_toggle_action_get_active (action);
  if (window->player) {
    swfdec_gtk_player_set_audio_enabled (SWFDEC_GTK_PLAYER (window->player),
	window->settings.sound);
  }
}

void
menu_file_close (GtkAction *action, SwfdecWindow *window)
{
  gtk_widget_destroy (window->window);
}

void
menu_help_about (GtkAction *action, SwfdecWindow *window)
{
  static const char *authors[] = {
    "Benjamin Otte <otte@gnome.org>",
    "Pekka Lampila <pekka.lampila@iki.fi>",
    NULL,
  };
  static const char *artists[] = {
    "Cristian Grada <krigenator@gmail.com>",
    NULL
  };

  gtk_show_about_dialog (NULL, 
      "logo-icon-name", "swfdec",
      "authors", authors,
      "artists", artists,
      "comments", "Play Adobe Flash files",
      "version", VERSION,
      "website", "http://swfdec.freedesktop.org/",
      NULL);
}

gboolean
main_window_destroy_cb (GtkWidget *widget, GdkEvent *event, SwfdecWindow *window)
{
  return FALSE;
}


thumbnailer/swfdec-thumbnailer.c
/* Swfdec Thumbnailer
 * Copyright (C) 2003, 2004 Bastien Nocera <hadess@hadess.net>
 *               2007 Pekka Lampila <pekka.lampila@iki.fi>
 *               2007 Benjamin Otte <otte@gnome.org>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA  02110-1301, USA.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <libswfdec/swfdec.h>

#define BORING_IMAGE_VARIANCE 256.0		/* Tweak this if necessary */

/* Copied from totem-video-thumbnailer.c and ported from GdkPixbuf to cairo */
/* This function attempts to detect images that are mostly solid images
 * It does this by calculating the statistical variance of the
 * black-and-white image */
static gboolean
is_image_interesting (cairo_surface_t *surface, int x, int y, int width,
    int height)
{
  /* We're gonna assume 8-bit samples. If anyone uses anything different,
   * it doesn't really matter cause it's gonna be ugly anyways */
  int stride, rowcount, num_samples, i, j;
  guchar *buffer;
  float x_bar, variance;

  g_return_val_if_fail (surface != NULL, FALSE);
  g_return_val_if_fail (x + width <= cairo_image_surface_get_width (surface),
      FALSE);
  g_return_val_if_fail (y + height <= cairo_image_surface_get_height (surface),
      FALSE);

  stride = cairo_image_surface_get_stride (surface);
  rowcount = cairo_image_surface_get_height (surface);
  buffer = cairo_image_surface_get_data (surface);

  x *= stride / cairo_image_surface_get_width (surface);
  width *= stride / cairo_image_surface_get_width (surface);

  num_samples = width * height;

  /* Iterate through the image to calculate x-bar */
  x_bar = 0.0f;
  for (i = y; i < y + height; i++) {
    for (j = x; j < x + width; j++) {
      x_bar += (float) buffer[i * stride + j];
    }
  }
  x_bar /= ((float) num_samples);

  /* Calculate the variance */
  variance = 0.0f;
  for (i = y; i < y + height; i++) {
    for (j = x; j < x + width; j++) {
      float tmp = ((float) buffer[i * stride + j] - x_bar);
      variance += tmp * tmp;
    }
  }
  variance /= ((float) (num_samples - 1));

  return (variance > BORING_IMAGE_VARIANCE);
}

static gboolean
advance (SwfdecPlayer *player, GTimer *timer, gulong advance_msecs)
{
  static const gulong max_runtime = 15 * 1000; /* 15 seconds per file */
  gulong elapsed;

  elapsed = (gulong) (g_timer_elapsed (timer, NULL) * 1000);
  if (elapsed >= max_runtime)
    return FALSE;
    
  swfdec_player_set_maximum_runtime (player, max_runtime - elapsed);
  swfdec_player_advance (player, advance_msecs);
  return TRUE;
}

int
main (int argc, char **argv)
{
  GOptionContext *context;
  GError *err;
  SwfdecPlayer *player;
  SwfdecLoader *loader;
  guint width, height;
  double scale, scaled_size, x, y, w, h;
  guint try;
  cairo_surface_t *surface;
  cairo_t *cr;
  int size = 128;
  char **filenames = NULL;
  GTimer *timer = NULL;
  gboolean time_left;
  const GOptionEntry entries[] = {
    {
      "size", 's', 0, G_OPTION_ARG_INT, &size,
      "Size of the thumbnail in pixels (default 128)", NULL
    },
    {
      G_OPTION_REMAINING, '\0', 0, G_OPTION_ARG_FILENAME_ARRAY, &filenames,
      NULL, "<INPUT FILE> <OUTPUT FILE>"
    },
    {
      NULL
    }
  };

  // init
  swfdec_init ();

  // read command line params
  context = g_option_context_new ("Create a thumbnail for Flash file");
  g_option_context_add_main_entries (context, entries, NULL);

  if (g_option_context_parse (context, &argc, &argv, &err) == FALSE) {
    g_printerr ("Couldn't parse command-line options: %s\n", err->message);
    g_error_free (err);
    return 1;
  }

  if (filenames == NULL || g_strv_length (filenames) != 2) {
    g_printerr ("One input and one output filename required\n");
    return 1;
  }

  loader = swfdec_file_loader_new (filenames[0]);
  if (loader->error) {
    g_printerr ("Error loading %s: %s\n", filenames[0], loader->error);
    g_object_unref (loader);
    return 1;
  }

  timer = g_timer_new ();
  player = swfdec_player_new (NULL);
  swfdec_player_set_loader (player, loader);
  g_timer_start (timer);

  /* Skip the first 10 seconds */
  /* Cheat: We do this in one call so that the max runtime kicks in */
  time_left = advance (player, timer, 10 * 1000);

  surface = cairo_image_surface_create (CAIRO_FORMAT_ARGB32, size, size);
  cr = cairo_create (surface);

  // get image size
  swfdec_player_get_default_size (player, &width, &height);
  if (width == 0 || height == 0) {
    /* force a size if the player doesn't have a default one */
    swfdec_player_set_size (player, size, size);
    width = height = size;
  }

  // determine amount of scaling to apply
  if (width > 2 * height) {
    scale = 0.5 * size / height;
  } else  if (height > 2 * width) {
    scale = 0.5 * size / width;
  } else {
    scale = (double) size / MAX (width, height);
  }

  // determine necessary translation to get image centered
  scaled_size = size / scale;
  x = (width - scaled_size) / 2;
  y = (height - scaled_size) / 2;
  cairo_scale (cr, scale, scale);
  cairo_translate (cr, -x, -y);

  // compute part to render
  x = MAX (x, 0);
  y = MAX (y, 0);
  w = MIN (width, scaled_size);
  h = MIN (height, scaled_size);

  // render the image
  swfdec_player_render (player, cr, x, y, w, h);

  for (try = 0; try < 10 && time_left; try++)
  {
    if (is_image_interesting (surface, (size - w * scale) / 2,
	  (size - h * scale) / 2, w * scale, h * scale))
	break;
    time_left = advance (player, timer, 1000);
    swfdec_player_render (player, cr, x, y, w, h);
  }

  cairo_destroy (cr);
  g_object_unref (player);
  g_timer_destroy (timer);

  cairo_surface_write_to_png (surface, filenames[1]);
  cairo_surface_destroy (surface);

  return 0;
}


player/swfdec-window.c
/* Swfdec Player
 * Copyright (C) 2007 Benjamin Otte <otte@gnome.org>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA  02110-1301, USA.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <glib/gi18n.h>
#include "swfdec-window.h"

G_DEFINE_TYPE (SwfdecWindow, swfdec_window, G_TYPE_OBJECT)

/* global list of windows */
static GSList *windows = NULL;
/* the default settings */
static const SwfdecWindowSettings default_settings = { FALSE, TRUE };

static void
swfdec_window_dispose (GObject *object)
{
  SwfdecWindow *window = SWFDEC_WINDOW (object);

  if (window->builder) {
    g_object_unref (window->builder);
    window->builder = NULL;
  }
  if (window->player) {
    g_object_unref (window->player);
    window->player = NULL;
  }

  G_OBJECT_CLASS (swfdec_window_parent_class)->dispose (object);
}

static void
swfdec_window_finalize (GObject *object)
{
  G_OBJECT_CLASS (swfdec_window_parent_class)->finalize (object);

  windows = g_slist_remove (windows, object);
  if (windows == NULL)
    gtk_main_quit ();
}

static void
swfdec_window_class_init (SwfdecWindowClass *klass)
{
  GObjectClass *object_class = G_OBJECT_CLASS (klass);

  object_class->dispose = swfdec_window_dispose;
  object_class->finalize = swfdec_window_finalize;
}

static void
swfdec_window_init (SwfdecWindow *window)
{
  window->settings = default_settings;

  windows = g_slist_prepend (windows, window);
}

/**
 * swfdec_window_set_url:
 * @window: the window that should show the given URL
 * @url: URL to show. Must be a valid file:// or http:// URL in UTF-8.
 *
 * Sets the URL of @window to be  @url, if no URL was set on @indow before.
 *
 * Returns: %TRUE if the URL could be set, %FALSE if the window already shows a 
 *          movie.
 **/
gboolean
swfdec_window_set_url (SwfdecWindow *window, const char *url)
{
  SwfdecWindowSettings settings;
  GObject *o;
  char *s;

  g_return_val_if_fail (SWFDEC_IS_WINDOW (window), FALSE);
  g_return_val_if_fail (url != NULL, FALSE);

  if (window->player || window->error)
    return FALSE;

  window->loader = swfdec_gtk_loader_new (url);
  window->player = swfdec_gtk_player_new (NULL);
  swfdec_player_set_loader (window->player, window->loader);
  swfdec_gtk_player_set_audio_enabled (SWFDEC_GTK_PLAYER (window->player), 
      window->settings.sound);
  o = gtk_builder_get_object (window->builder, "player-widget");
  swfdec_gtk_widget_set_player (SWFDEC_GTK_WIDGET (o), window->player);
  s = swfdec_loader_get_filename (window->loader);
  gtk_window_set_title (GTK_WINDOW (window->window), s);
  g_free (s);
  /* do this at the end to not get lag */
  swfdec_gtk_player_set_playing (SWFDEC_GTK_PLAYER (window->player), 
      window->settings.playing);

  return TRUE;
}

/**
 * swfdec_window_error:
 * @window: a window
 * @msg: an error message to display for the user
 *
 * Shows the given error message to the user and aborts playback.
 * This function may be called at any time, no matter the state of the window 
 * object.
 **/
void
swfdec_window_error (SwfdecWindow *window, const char *msg)
{
  /* NB: This function can be called during the construction process (like when
   * the UI file isn't found, so a window object may not even exist. */

  /* FIXME: disable playback related menu items */
  /* FIXME: output this in a saner way */
  g_printerr ("%s\n", msg);
  window->error = TRUE;
}

#define BUILDER_FILE DATADIR G_DIR_SEPARATOR_S "swfdec-gnome" G_DIR_SEPARATOR_S "swfdec-player.ui"
/**
 * swfdec_window_new:
 * @uri: a valid UTF-8 encoded URI.
 *
 * Opens a new window. If a URI was specified, it will be shown. Otherwise the
 * window will appear empty.
 *
 * Returns: The window that was just openend. You don't own a reference to it.
 **/
SwfdecWindow *
swfdec_window_new (const char *url)
{
  GError *error = NULL;
  SwfdecWindow *window;

  window = g_object_new (SWFDEC_TYPE_WINDOW, NULL);
  window->builder = gtk_builder_new ();
  gtk_builder_set_translation_domain (window->builder, GETTEXT_PACKAGE);
  if (!gtk_builder_add_from_file (window->builder, BUILDER_FILE, &error)) {
    swfdec_window_error (window, error->message);
    return window;
  }
  gtk_builder_connect_signals (window->builder, window);
  window->window = GTK_WIDGET (gtk_builder_get_object (window->builder, "player-window"));
  if (window->window == NULL) {
    swfdec_window_error (window, _("Internal error in the user interface definition"));
    return window;
  }
  g_object_weak_ref (G_OBJECT (window->window), (GWeakNotify) g_object_unref, window);
  if (url != NULL) {
    swfdec_window_set_url (window, url);
  }

  return window;
}

void
swfdec_window_set_settings (SwfdecWindow *window, const SwfdecWindowSettings *settings)
{
  SwfdecWindowSettings *org;

  g_return_if_fail (SWFDEC_IS_WINDOW (window));
  g_return_if_fail (settings != NULL);

  if (window->settings.playing != settings->playing) {
    GtkToggleAction *action = GTK_TOGGLE_ACTION (gtk_builder_get_object (window->builder, "play"));
    gtk_toggle_action_set_active (action, settings->playing);
    g_assert (window->settings.playing == settings->playing);
  }
  if (window->settings.sound != settings->sound) {
    GtkToggleAction *action = GTK_TOGGLE_ACTION (gtk_builder_get_object (window->builder, "mute"));
    gtk_toggle_action_set_active (action, !settings->sound);
    g_assert (window->settings.sound == settings->sound);
  }
}

